### PR TITLE
Document selectors for capabilities

### PR DESCRIPTION
**Pull request by toinehartman**
1. Set document selectors for capabilities, such that capabilities can only be triggered on documents that support them. Especially useful when registering multiple languages.
2. Dynamically register semantic highlighting capability (to test the former change, and make a start with registering all capabilities dynamically).
3. Instead of toggling the language for open editors, request a refresh of the open editors via LSP. This prevents races where requests can come in while the file is 'closed'.

This PR is a preparation to easily enable #1117.

### CODE REVIEW

File: rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/parametric/capabilities/CapabilityRegistration.java [hunk 1]
```diff
@@ -287,15 +290,42 @@ private <T> void handleError(@Nullable Throwable t, AbstractDynamicCapability<T>
...
+                        setDocumentSelector((TextDocumentRegistrationOptions) opts, providingLanguages);
+                    }
+                    return opts;
+                })
                 .thenApply(opts -> new Registration(cap.id(), cap.methodName(), opts));
         });
     }
 
+    private void setDocumentSelector(TextDocumentRegistrationOptions opts, Collection<ICapabilityParams> params) {
+        // If the document selector is set already by the capability implementation, we do not touch it.
```
Comment (DavyLandman) [inline rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/parametric/capabilities/CapabilityRegistration.java]: Can we explain why we do not touch it? as the condition in the if already says we do not touch it.

File: rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/parametric/ParametricTextDocumentService.java [hunk 2]
```diff
@@ -180,6 +183,8 @@ public class ParametricTextDocumentService extends TextDocumentStateManager impl
 
     private final String dedicatedLanguageName;
     private final SemanticTokenizer tokenizer = new SemanticTokenizer();
+    private final Set<String> extensionLessSchemes = new HashSet<>();
```
Comment (DavyLandman) [inline rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/parametric/ParametricTextDocumentService.java]: We will be writing to this from a single thread, but reading from different threads. That can cause messy problems.

To be safe we should use a concurrent hashset, (the trick with the keyset from the concurrent hashmap).

File: rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/parametric/ParametricTextDocumentService.java [hunk 3]
```diff
@@ -303,6 +308,21 @@ public void didOpen(DidOpenTextDocumentParams params) {
         TextDocumentState file = open(params.getTextDocument(), timestamp);
         handleParsingErrors(file, file.getCurrentDiagnosticsAsync());
         triggerAnalyzer(params.getTextDocument(), NORMAL_DEBOUNCE);
+
+        // Discover capabilities
+        discoverExtensionLessScheme(URIUtil.assumeCorrect(params.getTextDocument().getUri()));
```
Comment (DavyLandman) [inline rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/parametric/ParametricTextDocumentService.java:312]: Let's reuse the `file.getLocation()` here.

File: rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/parametric/ParametricTextDocumentService.java [hunk 4]
```diff
@@ -938,6 +953,32 @@ public synchronized void registerLanguage(LanguageParameter lang) {
                 refreshFileState(f);
             }
         }
+
+        // Do this last, after the parser for open editors has been updated,
+        // since this might trigger new requests on the editor contents.
+        updateCapabilities();
```
Comment (DavyLandman) [inline rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/parametric/ParametricTextDocumentService.java:954]: The old comment warned about `registerLanguage` and a race around that, why is the time when this is called still moved? is that race removed?

File: rascal-vscode-extension/src/lsp/ParameterizedLanguageServer.ts [hunk 5]
```diff
@@ -102,8 +102,7 @@ export class ParameterizedLanguageServer implements vscode.Disposable {
             for (const doc of vscode.workspace.textDocuments) {
                 const ext = path.extname(doc.uri.path);
                 if (ext !== "" && lang.extensions.includes(ext.substring(1))) {
-                    // (Re)set the language ID to re-trigger contribution requests
-                    await vscode.languages.setTextDocumentLanguage(doc, "plaintext");
+                    // Associate open editors with the given extensions with this language server
                     await vscode.languages.setTextDocumentLanguage(doc, this.languageId);
```
Comment (DavyLandman) [inline rascal-vscode-extension/src/lsp/ParameterizedLanguageServer.ts:106]: That `plaintext` work around was added for a reason, what was it, and why is it not valid anymore?

I'm sure we have no tests for this edge-case? Something like a file already open in VS Code when it starts, already assigned to parametric, but the language is not registered yet?

**Review by DavyLandman — CHANGES_REQUESTED** [review]
Great! I have some small concerns.

File: rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/parametric/ParametricTextDocumentService.java (hunk 3 above)
Comment (toinehartman) [reply rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/parametric/ParametricTextDocumentService.java:312]: We cannot, unfortunately, since if the opened file has an opaque URI (e.g. `untitled:Untitled-1`), `getLocation` will give a wrapped URI (`opaque-lsp-://untitled/Untitled-1`), and we lose the scheme. Of course, we could check whether `file.getLocation().getScheme()` is the opaque scheme (in which case we retrieve the original scheme from the authority), but that feels to dependent on the implementation of the wrapping of opaque URIs.

File: rascal-vscode-extension/src/lsp/ParameterizedLanguageServer.ts (hunk 5 above)
Comment (toinehartman) [reply rascal-vscode-extension/src/lsp/ParameterizedLanguageServer.ts:106]: It was added here: https://github.com/usethesource/rascal-language-servers/pull/809. I do not remember considering/discussing the `refresh*` requests, so we might just have missed those. I'll add a test for the case that you describe.

File: rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/parametric/ParametricTextDocumentService.java (hunk 2 above)
Comment (toinehartman) [reply rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/parametric/ParametricTextDocumentService.java]: I'll replace it with a thread-safe collection. I don't think we need to use the hashmap explicitly; there are [CopyOnWriteArraySet](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/concurrent/CopyOnWriteArraySet.html) and [ConcurrentSkipListSet](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/concurrent/ConcurrentSkipListSet.html).

File: rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/parametric/ParametricTextDocumentService.java (hunk 4 above)
Comment (toinehartman) [reply rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/parametric/ParametricTextDocumentService.java:954]: The race is not removed. It did not have to do with the order of operations within `registerLanguage`, but with multiple calls to `updateCapabilities` at the same time/from different threads. The comments have now been updated to reflect this.

File: rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/parametric/capabilities/CapabilityRegistration.java [hunk 6]
```diff
@@ -287,15 +290,43 @@ private <T> void handleError(@Nullable Throwable t, AbstractDynamicCapability<T>
...
+        // This allows specific capabilities to add specific selectors if needed.
```
Comment (DavyLandman) [inline rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/parametric/capabilities/CapabilityRegistration.java]: what if that capability forgets the scheme filters? is that okay?

**Review by DavyLandman — APPROVED** [review]

File: rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/parametric/capabilities/CapabilityRegistration.java (hunk 6 above)
Comment (toinehartman) [reply rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/parametric/capabilities/CapabilityRegistration.java]: Rewrote this to always add the extension- & scheme-based selectors to any selectors that the specific capability might have added. I am not yet aware of any capability that would need specific selectors that other capabilities do not need.

**Review by DavyLandman — APPROVED** [review]
